### PR TITLE
Reimplement Miunie.Discord with Discord.NET

### LINKED ISSUE #176: Phrase Based Command Handler
Create a handler in Miunie.Discord that hooks into message recieved event. 
This new handler would trugger services based on phrases. 
The command that should be integrated into this new system is the command: `"what do you think?"`

* (Ensure that a Mention Prefix @BotName is present)

### LINKED ISSUE #185: Reimplement Miunie.Discord with Discord.NET
fixes #176
relates to #110 (unblock)
relates to #186 (unblock)

The version of D#+ we are depending on is outdated and some features we're using are no longer available. Along with some other issues that need to be fixed, it is better to reimplement the Discord layer with Discord.NET.

### PR DESCRIPTION
**Pull request by petrspelos**
## Related Issue

fixes #185 
fixes #176
relates to #110 (unblock)
relates to #186 (unblock)

## Changes
The Miunie.Discord package has been rewritten to use Discord.NET as opposed to the non-existent version of D#+.

## Expected Feedback
Please make sure I'm not doing something insane. 😅 General sanity check and possible improvements are welcomed! ❤️

### CODE REVIEW

**Comment by DraxCodes** [comment]
bitmoji

Everything looks good, we have issues open for a couple things like the addition of a type reader, which would be the only improvement I could foresee being usefull, however the type reader can be done at a later date.

Only other thing I could think of was maybe to create our own ModuleBase, it'll make it super easy to migrate to a ShardedClient in future if needed. It also allows us to add some extension directly into `Context` would could be very nice to clean up with code in the modules. Again though, I think this is better suited to a new issue for a later date.

I will give this one more look tomorrow with some fresh eyes after I sleep. All being well, I shall approve and it can be them merged. 

Good job ma dude.
You are a real programmer.

File: src/Miunie.Discord/MiunieDiscord.cs [hunk 1]
```diff
@@ -1,21 +1,20 @@
...
 {
     public class MiunieDiscord : IMiunieDiscord
     {
         public string GetBotAvatarUrl()
-            => _discord.Client?.CurrentUser?.AvatarUrl;
+            => _discord.Client?.CurrentUser?.GetAvatarUrl();
 
-        private ConnectionState _connectionState;
-        public ConnectionState ConnectionState
+        private Core.ConnectionState _connectionState;
```
Comment (DraxCodes) [inline src/Miunie.Discord/MiunieDiscord.cs]: Maybe a rename here to something like MiunieConnectionState. Would make a little more sense seeing as it is Miunies connection state.

File: src/Miunie.Discord/MiunieDiscordClient.cs [hunk 2]
```diff
@@ -14,16 +15,14 @@ public MiunieDiscordClient(IBotConfiguration botConfig)
             _botConfig = botConfig;
         }
 
-        public void Initialize()
+        public async void Initialize()
```
Comment (DraxCodes) [inline src/Miunie.Discord/MiunieDiscordClient.cs]: Should be an async Task

File: src/nuget.config [hunk 3]
```diff
@@ -1,6 +1,6 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (DraxCodes) [inline src/nuget.config]: This file should no logger be needed.

File: src/Miunie.Discord/CommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            {
+                return;
+            }
+
+            var argPos = 0;
+            if (msg.HasMentionPrefix(_discord.Client.CurrentUser, ref argPos))
+            {
+                var context = new SocketCommandContext(_discord.Client, msg);
+                await TryRunAsBotCommand(context, argPos).ConfigureAwait(false);
+            }
```
Comment (svr333) [inline src/Miunie.Discord/CommandHandler.cs]: How about reverting the statement, if it doesn't have the prefix, then return?
You may want to do something else after the command has executed at some point.

File: src/Miunie.Discord/CommandModules/CurrencyCommand.cs [hunk 5]
```diff
@@ -19,28 +19,28 @@ public CurrencyCommand(EntityConvertor convertor, CurrencyService currencyServic
         }
 
         [Command("CZK")]
-        public async Task ShowCzkStatus(CommandContext ctx)
+        public async Task ShowCzkStatus()
```
Comment (svr333) [inline src/Miunie.Discord/CommandModules/CurrencyCommand.cs]: All these methods are async but they don't have the Async suffix, for consistency we should do that perhaps?

Comment (svr333) [reply src/Miunie.Discord/CommandModules/CurrencyCommand.cs]: Just checked the ProfileCommands and they do have the suffix, I'd say +1 to this then.

File: src/Miunie.Discord/CommandModules/ProfileCommand.cs [hunk 6]
```diff
@@ -1,54 +1,59 @@
...
             _entityConvertor = entityConvertor;
             _profileService = profileService;
+            _userConverter = userConverter;
         }
 
         [Command("profile")]
-        public async Task ShowProfileAsync(CommandContext ctx, MiunieUser m = null)
+        public async Task ShowProfileAsync(SocketGuildUser user = null)
         {
+            var m = _userConverter.DiscordMemberToMiunieUser(user ?? (Context.User as SocketGuildUser));
```
Comment (svr333) [inline src/Miunie.Discord/CommandModules/ProfileCommand.cs]: variable 'm' isn't very descriptive tbh - though I guess you know what is due to the context it is used in.

File: src/Miunie.Discord/Convertors/EntityConvertor.cs [hunk 7]
```diff
@@ -17,13 +17,13 @@ public EntityConvertor(IMiunieUserProvider miunieUserProvider)
             GuildConvertor = new MiunieGuildConvertor();
         }
 
-        public MiunieUser ConvertUser(DiscordMember m)
+        public MiunieUser ConvertUser(SocketGuildUser m)
```
Comment (svr333) [inline src/Miunie.Discord/Convertors/EntityConvertor.cs]: parameter named m doesn't make sense

File: src/Miunie.Discord/Convertors/MiunieGuildConvertor.cs [hunk 8]
```diff
@@ -1,22 +1,21 @@
...
 using Miunie.Core;
 using System.Linq;
 
 namespace Miunie.Discord.Convertors
 {
     public class MiunieGuildConvertor
     {
-        public MiunieGuild DiscordGuildToMiunieGuild(DiscordGuild g)
+        public MiunieGuild DiscordGuildToMiunieGuild(SocketGuild g)
             => new MiunieGuild
```
Comment (svr333) [inline src/Miunie.Discord/Convertors/MiunieGuildConvertor.cs]: Adding server_icon_url might be useful, though it has no usecase atm

File: src/Miunie.Discord/Convertors/MiunieUserConvertor.cs [hunk 9]
```diff
@@ -1,37 +1,27 @@
...
         }
 
-        public async Task<Optional<MiunieUser>> ConvertAsync(string userInput, CommandContext context)
-        {
-            var result = await _dmConverter.ConvertAsync(userInput, context);
-            return DiscordMemberToMiunieUser(result.Value);
-        }
-
-        public MiunieUser DiscordMemberToMiunieUser(DiscordMember user)
+        public MiunieUser DiscordMemberToMiunieUser(SocketGuildUser user)
```
[the remark this answers is not in the document]
Comment (svr333) [reply src/Miunie.Discord/Convertors/MiunieUserConvertor.cs]: Could be a good idea, though I don't see a big problem like it is now

File: src/Miunie.Discord/Logging/DiscordLogger.cs [hunk 10]
```diff
@@ -13,16 +13,18 @@ public DiscordLogger(ILogWriter logger)
             _logger = logger;
         }
 
-        internal void Log(object sender, DebugLogMessageEventArgs e)
+        internal Task Log(LogMessage evt)
```
Comment (svr333) [inline src/Miunie.Discord/Logging/DiscordLogger.cs]: > `(LogMessage evt)`

What does evt stand for? - isn't `(LogMessage log)` better in this context?

File: src/Miunie.Discord/MiunieDiscordClient.cs [hunk 11]
```diff
@@ -14,16 +15,14 @@ public MiunieDiscordClient(IBotConfiguration botConfig)
...
         {
-            Client = new DiscordClient(new DiscordConfiguration
+            Client = new DiscordSocketClient(new DiscordSocketConfig
             {
-                Token = _botConfig.DiscordToken,
-                TokenType = TokenType.Bot,
-                AutoReconnect = true,
-                LogLevel = LogLevel.Info,
-                UseInternalLogHandler = false
+                LogLevel = LogSeverity.Info,
```
Comment (svr333) [inline src/Miunie.Discord/MiunieDiscordClient.cs]: I also like to mention other stuff even though it defaults to that, it just makes it more clear when reading - though, that's complete personal preference and also bloat

File: src/Miunie.Discord/Adapters/DiscordGuildsAdapter.cs [hunk 12]
```diff
@@ -16,11 +16,11 @@ public DiscordGuildsAdapter(IDiscord discord, EntityConvertor entityConvertor)
             _entityConvertor = entityConvertor;
         }
 
-        public async Task<MiunieGuild> FromAsync(MiunieUser user)
+        public Task<MiunieGuild> FromAsync(MiunieUser user)
```
Comment (svr333) [inline src/Miunie.Discord/Adapters/DiscordGuildsAdapter.cs]: Method isn't async but has the suffix Async, it is a Task though so 🤷‍♂

Comment (DraxCodes) [reply src/Miunie.Discord/Adapters/DiscordGuildsAdapter.cs]: Async suffix is used to display that the method can be run asynchronously. FromAsync is the correct format.

File: src/Miunie.Discord/Logging/DiscordLogger.cs (hunk 10 above)
Comment (DraxCodes) [reply src/Miunie.Discord/Logging/DiscordLogger.cs]: Either that or args.

File: src/Miunie.Discord/Convertors/MiunieUserConvertor.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (DraxCodes) [reply src/Miunie.Discord/Convertors/MiunieUserConvertor.cs]: The only issue is the inconsistency. One Converter is an extension method, the others are injected through the IOC container.

**Comment by svr333** [comment]
Did you really?

**Pull request merged**